### ISSUE TITLE
Compiler incorrectly warns about type mismatch

### ISSUE THREAD

**Issue by vobruba-martin** [google/closure-compiler#3570]
```javascript
/**
 * @param {?Blob} blobOrNull
 */
function fn(blobOrNull) {
	/** @type {!Blob} */
	var blob;

	if (blobOrNull) {
		blob = blobOrNull;
	}
	(function() {
		blobOrNull = null;
	})();
}
```
```
test.js:9: WARNING - [JSC_TYPE_MISMATCH] assignment
found   : (Blob|null)
required: Blob
		blob = blobOrNull;
		^^^^^^^^^^^^^^^^^

0 error(s), 1 warning(s), 85.7% typed
```

But no warning here:
```javascript
/**
 * @param {?Blob} blobOrNull
 */
function fn(blobOrNull) {
	/** @type {!Blob} */
	var blob;

	if (blobOrNull) {
		blob = blobOrNull;
	}
	blobOrNull = null;
}
```
```
0 error(s), 0 warning(s), 100.0% typed
```

**Comment by evmar** [comment]
The compiler assumes that it cannot predict when the code within nested functions run.  We could maybe do something smarter in the special case of an IIFE, but it is unlikely we'll do this.
